lib/diag.c

```diff
@@ -907,16 +911,180 @@ int switchtec_diag_refclk_ctl(struct switchtec_dev *dev, int stack_id, bool en)
+	status.port = port;
+	ret = switchtec_cmd(dev, MRPC_DIAG_PORT_LTSSM_LOG, &status,
+			    sizeof(status), &status_output,
+			    sizeof(status_output));
+	if (ret)
+		return ret;
+
+	*log_count = status_output.log_count;
+
+	/* get log data */
+	log_dump.sub_cmd = MRPC_LTMON_LOG_DUMP_GEN5;
+	log_dump.port = port;
+	log_dump.log_index = 0;
+	log_dump.no_of_logs = *log_count;
+
+	if(log_dump.no_of_logs <= SWITCHTEC_LTSSM_MAX_LOGS) {
+		/* Single buffer log case */
+		ret = switchtec_cmd(dev, MRPC_DIAG_PORT_LTSSM_LOG, &log_dump,
+				    sizeof(log_dump), &log_buffer[0], 
+				    log_dump.no_of_logs * sizeof(struct log_dump_out) + 4);
+		if (ret)
+			return ret;
+		for (int i = 0; i < log_dump.no_of_logs; i++) {
+			dw0 = log_dump_out_ptr[i].dw0;
+			timestamp = log_dump_out_ptr[i].ram_timestamp;
+	
+			rate = (dw0 >> 13) & SWITCHTEC_LTSSM_LOG_MASK_RATE;
+			major = (dw0 >> 7) & SWITCHTEC_LTSSM_LOG_MASK_MAJOR;
+			minor = (dw0 >> 3) & SWITCHTEC_LTSSM_LOG_MASK_MINOR;
+	
+			log_data[i].timestamp = timestamp;
+			log_data[i].link_rate = switchtec_gen_transfers[rate+1];
+			log_data[i].link_state = major | (minor << 8);
+		}
+	} else {
+		/* Multiple buffer log case */
+		int buff_count = log_dump.no_of_logs / SWITCHTEC_LTSSM_MAX_LOGS;
+		int curr_idx = 0;
+		int buffer_size = SWITCHTEC_LTSSM_MAX_LOGS * sizeof(struct log_dump_out) + 4;
+
+		for (int i = 0; i < buff_count; i++) {
+			log_dump.no_of_logs = SWITCHTEC_LTSSM_MAX_LOGS;
+			ret = switchtec_cmd(dev, MRPC_DIAG_PORT_LTSSM_LOG, &log_dump,
+					    sizeof(log_dump), &log_buffer[0], 
+					    buffer_size);
+			if (ret)
+				return ret;
+			log_dump_out_ptr = (struct log_dump_out *)&(log_buffer[4]);
+			for (int j = 0; j < log_dump.no_of_logs; j++) {
+				dw0 = log_dump_out_ptr[j].dw0;
+				timestamp = log_dump_out_ptr[j].ram_timestamp;
+		
+				rate = (dw0 >> 13) & SWITCHTEC_LTSSM_LOG_MASK_RATE;
+				major = (dw0 >> 7) & SWITCHTEC_LTSSM_LOG_MASK_MAJOR;
+				minor = (dw0 >> 3) & SWITCHTEC_LTSSM_LOG_MASK_MINOR;
+		
+				log_data[curr_idx + j].timestamp = timestamp;
+				log_data[curr_idx + j].link_rate = switchtec_gen_transfers[rate+1];
+				log_data[curr_idx + j].link_state = major | (minor << 8);
+			}
+			curr_idx += SWITCHTEC_LTSSM_MAX_LOGS;
+			log_dump.log_index = curr_idx;
+		}
+		if (*log_count % SWITCHTEC_LTSSM_MAX_LOGS) {
+			log_dump.no_of_logs = *log_count - curr_idx;
+			ret = switchtec_cmd(dev, MRPC_DIAG_PORT_LTSSM_LOG, &log_dump,
+					    sizeof(log_dump), &log_buffer[0], 
+					    log_dump.no_of_logs * sizeof(struct log_dump_out) + 4);
+			if (ret)
+				return ret;
+			log_dump_out_ptr = (struct log_dump_out *)&(log_buffer[4]);
+			for (int i = 0; i < log_dump.no_of_logs; i++) {
+				dw0 = log_dump_out_ptr[i].dw0;
+				timestamp = log_dump_out_ptr[i].ram_timestamp;
+		
+				rate = (dw0 >> 13) & SWITCHTEC_LTSSM_LOG_MASK_RATE;
+				major = (dw0 >> 7) & SWITCHTEC_LTSSM_LOG_MASK_MAJOR;
+				minor = (dw0 >> 3) & SWITCHTEC_LTSSM_LOG_MASK_MINOR;
```

Comment:
If a helper function is possible, we probably can remove the mask macros and just expand the value in the statement, just as we put shift offset values here.